### ISSUE TITLE
Run tile should be more informative

### ISSUE THREAD

**Issue by adriatic** [monterey-framework/monterey#334]
### Intention

**I'm submitting a feature request**  

![image](https://cloud.githubusercontent.com/assets/2712405/18730410/34c2eaa0-8024-11e6-8e13-b8ba7df8cf4e.png)

I believe that it would be cool to change the glyph from the triangle to a yellow square, that should subsequently turn into a green square when app is loaded. As things are now, click on `Run`, followed by a click on the browser tile shows the `This site can’t be reached` error in the browser.

**Comment by JeroenVinke** [comment]
While I like the idea, for the implementation we would know when the workflow is done but not finished (the web server started but the gulp watch process is still running). For the Run workflow we do know this, but for custom workflows (the ones the user will create in addition to the run workflow) we do not. So the indicator would stay yellow even though some is done already. So this is not the best solution

I will change the behavior so that the taskmanager is opened when starting a workflow so tha the user can monitor the progress of the task and click the URL to open the application. I will create a setting in the preferences so you can decide whether or not to show the taskmanager when starting a workflow

**Comment by adriatic** [comment]
The fix "show taskmanager when starting workflow" seems like the most obvious choice. In order to make absolutely precise, I would label this change as "show taskmanager when clicking on the Run tile" as this was the scenario in which we started this discussion.

I would also like to comment on few statements from @JeroenVinke's post above:

> For the Run workflow we do know this to some degree

What means "some degree"? - Monterey is fully aware what is going on, proven by the fact that it will show the application URL when `gulp watch` reached the required state.

> for custom workflows (the ones users will create in addition to the run workflow) we do not (know when "ready")

Can you please explain this with more details? It seems as if you created a design for "custom workflows" that I know **nothing** about, where I would really like to know **everything** about 😏

**Comment by JeroenVinke** [comment]
> What means "some degree"?

Since we know what to look for in the output of certain commands, we can often detect that something has happened. Take a look at https://github.com/monterey-framework/monterey/issues/309#issuecomment-247108455 where the output of the `gulp watch` command of a project was different than what we expected, causing monterey not to know what URL the project is running under. If you decide to use a different web server than what the skeleton uses, or the output of a command is a little bit different than expected, then this detection will fail, so I see this as unreliable (but still very useful)

> Can you please explain this with more details? It seems as if you created a design for "custom workflows" that I know nothing about, where I would really like to know everything about 😏

A workflow is the definition of a set of commands and every workflow has a name (such as "Run" or "Deploy" or something else). When a workflow is started, the commands defined in the workflow areto executed in a particular order, so you can see it as a tree of commands. 

When you create a `skeleton-esnext` project, Monterey will create a workflow called `Run` with the following "tree" of commands:

-- gulp watch

Note that this is just one command, but see this as a tree nonetheless. 

If you create a `skeleton-typescript-apsnetcore` project, Monterey creates a workflow called `Run` with this tree of commands:

-- gulp watch
-- dotnet run

Note that these commands are at the same level in the tree, causing them to be started at the same time (there is no dependency between the two).

If you were to create your own workflow (lets call it the "Deploy" workflow):

-- gulp build
&nbsp;&nbsp;&nbsp;&nbsp;-- gulp deploy

then it means that you want the `gulp build` command to finish executing before starting `gulp deploy`.

The `Run` tile we have now is a tile for a workflow with the name `Run`. If you were to create a workflow called `Deploy` then you would see a tile called `Deploy` too that would work in the same way as the `Run` tile.

When clicking on such tile, all commands in the workflow definition are added to the TaskManager. You could say that the Run tile does not know how to start the application, it only knows that it must start `gulp watch` when clicked.

Right now we have only the `Run` workflow, but later users could compose workflows themselves). Given the complexity of the UI we decided to hide this "Workflow editor" tile for now: https://github.com/monterey-framework/monterey/issues/323

![image](https://cloud.githubusercontent.com/assets/2189477/18750038/bdd440d6-80d9-11e6-80dc-34f1927ceaf0.png)

**Comment by adriatic** [comment]
> The `Run` tile we have now is a tile for a workflow with the name `Run`.

This is where I got lost - as I was under the impression that `Run` is completely generic (run whatever workflow is being defined so far). I also remain pretty sure that this would be a better definition for the `Run` tile. In this case explanation like

> You could say that the `Run` tile does not know how to start the application, it only knows that it must start `gulp watch` when clicked.

would not be accurate and thus not needed. Since I know that you know how to think straight, you must have some reasons to make it different from what feels "normal" to me.

Thank you for such nice explanation - all such write-ups will come really handy for our documentation efforts, of course.

**Comment by JeroenVinke** [comment]
> This is where I got lost - as I was under the impression that Run is completely generic (run whatever workflow is being defined so far)

Let me say it this way, if you were to rename the `Run` workflow to `Start app` then you would end up with a tile called `Start app`. 

![image](https://cloud.githubusercontent.com/assets/2189477/18751600/e2295e8e-80df-11e6-8926-bf9ccd6bae6a.png)

In the above sreenshot there are two workflows: "Start app" (which runs `gulp watch`) and "Deploy` (which runs `gulp deploy`).

**Comment by adriatic** [comment]
In my view of Monterey from the day one (which I know that it matches yours as well), `Workflow` is the center piece, despite the humble beginning, where there was a panel in the wizard that would show just the list of commands like

```
npm install
jspm install
gulp watch
```

I insisted in calling this `Workflow` - at least in the documentation which I wrote in parallel with Monterey development. In addition to the concept of tiles, which supports the concept of extensibility, it is the `Workflow` concept that adds programmability. These two concepts are key to Monterey's success and we all know that.

So, I disagree with your opinion stated as

> Let me say it this way, if you were to rename the `Run` workflow to `Start app` then you would end up with a tile called `Start app`. 

`Run` is a shorthand for `Run workflow` - and this is a completely well defined situation, as even if the user forgot what the workflow will do, he/she can click on the Workflow tile and find out.

In our current situation (where there is no Workflow tile yet), click on the `Run` tile has still the same semantic - `Run the damn workflow`, regardless of the possible perception of few users that would think that we are pretentious (calling a sequence of commands "workflow").

**Comment by JeroenVinke** [comment]
So you don't want users to be able to create their own workflows? So for example if you run the same three commands each day it isn't possible to create a tile for this?

**Comment by adriatic** [comment]
Well, it seems that I am off by a foot somewhere, as I do not see my statement leading to such conclusion - which comes as a consequence of the fact that I am way behind your concept of what Monterey workflow really is (note that I am not implying anything else but my ignorance, at the moment 😏 )

I believe that you started the work on workflow design based on my use of that word in the context of the definition of the `set of commands needed to build and run` something (**build a library** - as pertinent to the type of work that Rob does every day, **build and run some application**, etc; we never discussed the concept of deployment as this is the key portion of the Docker integration with Monterey - design of which has yet to start).

Since we are not showing the workflow tile, this discussion is a few days premature, as we both want to release this current "beast". So, for the record: of course I want to be able to provide the ability to declare `aliases` for a workflow, this current discussion only shows that we did not sufficiently discuss the workflow implementation - a fact that we both know 😄 

As it might be good to clarify my view of my role in this whole "circus", let me do that as well: I want to participate in "Monterey everything", design included - not as a dictator, but as a participant that will seek consensus - just like other team members expect. Due to my experience, I might beat on something more forcefully than I would if I had no idea what is being discussed - but in that case I am pretty confident that other team members will tell me that directly 😏

**Comment by JeroenVinke** [comment]
> Well, it seems that I am off by a foot somewhere, as I do not see my statement leading to such conclusion - which comes as a consequence of the fact that I am way behind your concept of what Monterey workflow really is (note that I am not implying anything else but my ignorance, at the moment 😏 )

This makes sense, so we should figure out whether or not we are on the same page and I believe that understanding the current implementation is key.

> I believe that you started the work on workflow design based on my use of that word in the context of the definition of the set of commands needed to build and run something (build a library - as pertinent to the type of work that Rob does every day, build and run some application, etc; we never discussed the concept of deployment as this is the key portion of the Docker integration with Monterey - design of which has yet to start).

I am not proposing to add the `Deploy` workflow/tile. I am trying to explain how the current implementation is, so that we can use that as a starting point for further discussions. I will give it another shot 😄 

As currently implemented, every workflow has two things: a name and a set of commands. The following are all valid workflows:

**Run**
Name: Run
Commands: ['gulp watch']

**Build**
Name: Build
Commands: ['gulp build-css', 'gulp build-typescript']

**Reinstall JSPM packages**
Name: Reinstall JSPM packages
Commands: ['jspm cc', 'rmdir jspm_packages', 'jspm install']

These are all valid workflow definitions. I could give you a million of these examples just to indicate that a workflow is a collection of **any** commands. Right now we only have the Run workflow, which has commands that start an application.

In order to start workflows, we have tiles.

![image](https://cloud.githubusercontent.com/assets/2189477/18756347/e4423e24-80ef-11e6-8798-a7c3d520a3c2.png)

**Because we only have one workflow at the moment (and the user cannot add their own workflows yet) there is only one tile that looks like this**.

The text in the top part of the tile reflects the name of the workflow. The play/stop indicator displays whether or not the workflow is currently running or not. If you would have 3 workflows, then you would have 3 separate tiles that can all be started and stopped independently.

A very simplified version of all this: **when you click on a tile with a start/stop indicator (a workflow tile), Monterey will execute the commands that are defined in the workflow.**

> As it might be good to clarify my view of my role in this whole "circus", let me do that as well: I want to participate in "Monterey everything", design included - not as a dictator, but as a participant that will seek consensus - just like other team members expect. Due to my experience, I might beat on something more forcefully than I would if I had no idea what is being discussed - but in that case I am pretty confident that other team members will tell me that directly 

I am not sure why this needed to be reiterated, I believe https://github.com/monterey-framework/monterey/issues/284#issue-175734165 was pretty clear on what problem I thought "workflows" would solve, and I would have expected you to hit the breaks if that was far away from what you had in mind 😄 I definitely did not exclude you from the design process

**Comment by adriatic** [comment]
First thing first : we are pretty much in sync with everything; the only difference that I could see is that you see a workflow as a "set of commands", where is see it as a "structured set of commands and verbs" - a mini language, that would eventually be able to have loops and conditional commands. At this time, it is not important to discuss these issues.

Then, there is one (probably just created) misunderstanding - as you write:

> I would by no means call this "docker integration"

So, you must have incorrectly interpreted my view that means to say that the deployment tasks will be best implemented once we add Docker - see the screenshot below:

![image](https://cloud.githubusercontent.com/assets/2712405/18760926/01724602-80d1-11e6-92aa-ed487e6f2c3f.png)

I might be exceeding your expectations of what a workflow could be - as I see the possibility where we have a workflow engine, an extensive library of workflows with the workflow editor that allows the users to create their own workflows (making the library extensible as well) and the support for assigning workflows to projects. Note my cautious approach: **`I see the possibility where ...`** - leaving all this to future discussions. 

In summary: my "intervention" was meant to ensure alignment between what you believe that we should do, and what I believe we should do. This of course includes other team members (current and future) - they should pitch in at any time ...

**Comment by JeroenVinke** [comment]
> First thing first : we are pretty much in sync with everything; the only difference that I could see is that you see a workflow as a "set of commands", where is see it as a "structured set of commands and verbs" - a mini language, that would eventually be able to have loops and conditional commands. At this time, it is not important to discuss these issues.

Good 👍 

>Then, there is one (probably just created) misunderstanding - as you write:
> >I would by no means call this "docker integration"

No need to discuss this any further. I was using the Deploy workflow as an example for what a workflow **could** do, not as a proposal to actually create the deploy workflow. Deployment is something we still need to start the design work on, yes

> I might be exceeding your expectations of what a workflow could be - as I see the possibility where we have a workflow engine, an extensive library of workflows with the workflow editor that allows the users to create their own workflows (making the library extensible as well) and the support for assigning workflows to projects. Note my cautious approach: I see the possibility where ... - leaving all this to future discussions.

Understood, and I will definitely point back to this issue when we get the "signal" for us to improve/extend the current implementation to something more advanced

**Comment by adriatic** [comment]
Democracy at its best 😄

**Comment by JeroenVinke** [comment]
Closing this issue as we fixed the problem and the info in here is still available when closed
